api/src/main/java/com/github/eyefloaters/console/api/service/TopicService.java

```diff
@@ -376,4 +391,56 @@ Either<OffsetInfo, Throwable> either(ListOffsetsResultInfo result, Throwable err
 
         return Either.of(result, error, transformer);
     }
+
+    CompletionStage<Void> describeLogDirs(Admin adminClient, Map<Uuid, Either<Topic, Throwable>> topics) {
+        Map<String, Uuid> topicIds = new HashMap<>(topics.size());
+
+        var topicPartitionReplicas = topicPartitionReplicas(topics, topicIds);
+        var nodeIds = topicPartitionReplicas.values().stream().flatMap(Collection::stream).distinct().toList();
+        var logDirs = adminClient.describeLogDirs(nodeIds, new DescribeLogDirsOptions()
+                .timeoutMs(5000))
+                .descriptions();
+        var promise = new CompletableFuture<Void>();
+
+        var pendingInfo = topicPartitionReplicas.entrySet()
+            .stream()
+            .flatMap(e -> e.getValue().stream().map(node -> Map.entry(e.getKey(), node)))
+            .map(e -> {
+                var topicPartition = e.getKey().toKafkaModel();
+                int nodeId = e.getValue();
+                var partitionInfo = topics.get(topicIds.get(topicPartition.topic()))
+                        .getPrimary()
+                        .getPartitions()
+                        .getPrimary()
+                        .stream()
+                        .filter(p -> p.getPartition() == topicPartition.partition())
+                        .findFirst();
+
+                return logDirs.get(nodeId).toCompletionStage().<Void>handle((nodeLogDirs, error) -> {
+                    if (error != null) {
+                        partitionInfo.ifPresent(p -> p.setReplicaLocalStorage(nodeId, Either.ofAlternate(error)));
+                    } else {
+                        nodeLogDirs.values()
+                            .stream()
+                            .map(dir -> dir.replicaInfos())
+                            .map(replicas -> replicas.get(topicPartition))
+                            .filter(Objects::nonNull)
+                            .map(org.apache.kafka.clients.admin.ReplicaInfo.class::cast)
+                            .map(ReplicaLocalStorage::fromKafkaModel)
+                            .forEach(replicaInfo -> partitionInfo.ifPresent(p -> p.setReplicaLocalStorage(nodeId, Either.of(replicaInfo))));
+                    }
+
+                    return null;
+                });
+            })
+            .map(CompletionStage::toCompletableFuture)
+            .toArray(CompletableFuture[]::new);
+
+        CompletableFuture.allOf(pendingInfo)
+            .thenApply(nothing -> promise.complete(null))
+            .exceptionally(promise::completeExceptionally);
```

Comment:
This pattern again

Reply:
This one can definitely be simplified :+1: